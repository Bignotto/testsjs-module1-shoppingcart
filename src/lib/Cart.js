import { remove, find } from 'lodash';
import Dinero from 'dinero.js';

const Money = Dinero;
Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

import { calculateDiscount } from './discount.utils';

export default class Cart {
  items = [];

  add(item) {
    if (find(this.items, { product: item.product }))
      remove(this.items, { product: item.product });
    this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product });
  }

  summary() {
    const total = this.getTotal().getAmount();
    const items = this.items;
    const formattedTotal = this.getTotal().toFormat('$0,0.00');

    return {
      total,
      formattedTotal,
      items,
    };
  }

  checkout() {
    const { total, items } = this.summary();

    this.items = [];

    return {
      total,
      items,
    };
  }

  getTotal() {
    return this.items.reduce((acc, { quantity, product, condition }) => {
      const amount = Money({ amount: quantity * product.price });
      let discount = Money({ amount: 0 });

      if (condition) {
        discount = calculateDiscount(amount, quantity, condition);
      }

      return acc.add(amount.subtract(discount));
    }, Money({ amount: 0 }));
  }
}
