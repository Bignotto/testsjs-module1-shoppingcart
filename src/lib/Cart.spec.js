import Cart from './Cart';

describe('Cart', () => {
  let cart;
  let product;
  let product2;

  beforeEach(() => {
    cart = new Cart();
    product = {
      title: 'Fraldinha Red Angus',
      price: 6875,
    };
    product2 = {
      title: 'Entranha Fina',
      price: 11990,
    };
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created cart', () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it('should multiply quantity and price and calculate the total amount', () => {
      const item = {
        product,
        quantity: 2, //70776
      };
      cart.add(item);

      expect(cart.getTotal().getAmount()).toEqual(13750);
    });

    it('should ensure only one of each product in shopping cart', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });
      expect(cart.getTotal().getAmount()).toEqual(6875);
    });

    it('should update total when a product gets included then removed', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toEqual(11990);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and a list of items', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should return an object with the total and a list of items when sumary() is calle', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it('should return formatted amount when summary is called', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.summary().formattedTotal).toEqual('R$257.40');
    });

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.checkout();

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });

  describe('special conditions', () => {
    it('should apply a discount over the item total when a minimum quantity threshold is passed', () => {
      const condition = {
        percentage: 30,
        minimum: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 3,
      });

      expect(cart.getTotal().getAmount()).toEqual(14437);
    });

    it('should apply a discount over the item total for even quantities', () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 4,
      });

      expect(cart.getTotal().getAmount()).toEqual(13750);
    });

    it('should apply a discount over the item total for odd quantities', () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(20625);
    });

    //
    it('should not apply a discount over the item total when a minimum quantity threshold is not passed', () => {
      const condition = {
        percentage: 30,
        minimum: 4,
      };

      cart.add({
        product,
        condition,
        quantity: 3,
      });

      expect(cart.getTotal().getAmount()).toEqual(20625);
    });

    it('should not apply a discount over the item total for even quantities', () => {
      const condition = {
        quantity: 6,
      };

      cart.add({
        product,
        condition,
        quantity: 4,
      });

      expect(cart.getTotal().getAmount()).toEqual(27500);
    });

    it('should receive two or more conditions and apply the one with higher discount case 1', () => {
      const condition1 = {
        quantity: 2,
      };
      const condition2 = {
        percentage: 30,
        minimum: 2,
      };

      cart.add({
        product,
        condition: [condition1, condition2],
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(20625);
    });

    it('should receive two or more conditions and apply the one with higher discount case 2', () => {
      const condition1 = {
        quantity: 2,
      };
      const condition2 = {
        percentage: 80,
        minimum: 2,
      };

      cart.add({
        product,
        condition: [condition1, condition2],
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(6875);
    });
  });
});
