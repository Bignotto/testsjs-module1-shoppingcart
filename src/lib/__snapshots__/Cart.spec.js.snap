// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and a list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 6875,
        "title": "Fraldinha Red Angus",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 11990,
        "title": "Entranha Fina",
      },
      "quantity": 1,
    },
  ],
  "total": 25740,
}
`;

exports[`Cart checkout() should return an object with the total and a list of items when sumary() is calle 1`] = `
Object {
  "formattedTotal": "R$257.40",
  "items": Array [
    Object {
      "product": Object {
        "price": 6875,
        "title": "Fraldinha Red Angus",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 11990,
        "title": "Entranha Fina",
      },
      "quantity": 1,
    },
  ],
  "total": 25740,
}
`;
